pkg/client/mapper.go

```diff
@@ -103,3 +103,57 @@ func NewBidirectionalNewOneToManyGroupMapper(source, dest, groupMappingFile stri
 	}
 	return nil, nil, fmt.Errorf("unsupported source to dest mapper type: source %s, dest %s", source, dest)
 }
+
+// UserMapperImpl implements groupsync.UserMapper.
+type UserMapperImpl map[string]string
+
+func (u UserMapperImpl) MappedUserID(ctx context.Context, userID string) (string, error) {
+	v, ok := u[userID]
+	if !ok {
+		return "", groupsync.ErrTargetUserIDNotFound
+	}
+	return v, nil
+}
+
+// NewGoogleGroupGitHubUserMapper creates a UserMapperImpl that maps
+// google user email to github user handle.
+func NewGoogleGroupGitHubUserMapper(userMappingFile string) (groupsync.UserMapper, error) {
+	b, err := os.ReadFile(userMappingFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read mapping file: %w", err)
+	}
+	tm := &v1alpha3.UserMappings{}
+	if err := prototext.Unmarshal(b, tm); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal mapping file: %w", err)
+	}
+
+	ggToGHUserMapping := make(UserMapperImpl)
+	ghToGGUserMapping := make(UserMapperImpl)
+
+	for _, mapping := range tm.GetMappings() {
+		src, dst := mapping.GetGoogleUserEmail(), mapping.GetGitHubUserId()
+		// skip user if he doesn't have google group or github that needs mappings.
+		if src == "" || dst == "" {
+			continue
+		}
+		// Check user mapping relation is 1:1.
+		if existingDst, ok := ggToGHUserMapping[src]; ok && existingDst != dst {
+			return nil, fmt.Errorf("google group user %s mapped to multiple github user %s,%s", src, existingDst, dst)
+		}
+		ggToGHUserMapping[src] = dst
+
+		if existingSrc, ok := ghToGGUserMapping[dst]; ok && existingSrc != src {
+			return nil, fmt.Errorf("github user %s mapped to multiple google group user %s,%s", dst, existingSrc, src)
+		}
+		ghToGGUserMapping[dst] = src
+	}
+	return ggToGHUserMapping, nil
+}
+
+// NewUserMapperImpl creats a UserMapperImpl base on source and dest system type.
+func NewUserMapperImpl(source, dest, mappingFilePath string) (groupsync.UserMapper, error) {
+	if source == tltypes.SystemTypeGoogleGroups && dest == tltypes.SystemTypeGitHub {
+		return NewGoogleGroupGitHubUserMapper(mappingFilePath)
+	}
+	return nil, fmt.Errorf("unsupported source to dest user mapper type: source %s, dest %s", source, dest)
+}
```

Comment:
For now, we are implementing this to support group sync from GoogleGroups to Github, and the next step is support from GoogleGroup to GitLab. So I am planning on using the source and dest system type to create different mapper.

This code will be built as a binary and this same binary can be ran by both github workflow and gitlab

Reply:
Wouldn't that be a different proto and therefore a different structure? It feels weird to force a generic interface here, when you're going to need to switch on the proto message anyway further up the stack.

Reply:
When added GitLab user mapping support, we just need to add 'optional string git_lab_user_id = 3` to `UserMapping` proto, and in `NewGoogleGroupGitLabUserMapper`, we read the same text proto file,  create a `UserMapper` that maps from Google Email to Gitlab user name. 
It will be the same proto but with added fields. And this way if users can still use the same textproto file even the binary is updated with more supported system type.

Reply:
I'm not sure that's the right pattern; that's really overloading the proto, especially once you have github-specific or gitlab-specific information. It probably wants to be a `OneOf`, which handles this switching anway.

Reply:
Discussed offline, cleared confusions and mis-communications, now using `NewUserMapper` to call constructor for different src->dst systems.